api/controllers/v0.3/ILSHelper.js

```diff
@@ -0,0 +1,390 @@
+
+  constructor(ilsClient = undefined) {
+    this.client =
+      ilsClient ||
+      new ILSAPIClient({
+        wsdl: this.ilsUrl,
+        open_timeout: 15,
+        read_timeout: 15,
+        log: true,
+        filters: ["username", "password"],
+        log_level: "debug",
+        pretty_print_xml: true,
+      });
+  }
+
+  ilsUrl() {
+    return `https://${process.env["ILS_SERVER"]}/iii/patronio/services/PatronIO?wsdl`;
+  }
+
+  callAndFailGracefully(params) {
+    try {
+      this.client.call(params);
+    } catch (e) {
+      let error;
+      switch (
+        e["fault"]["detail"]["com_iii_webapps_ws_patronio_patron_io_fault"][
+          "code"
+        ]
+      ) {
+        case IlsHelper.NOT_FOUND:
+          error = "NotFoundError()";
+          break;
+        case IlsHelper.MULTIPLE_MATCHES:
+          error = "MultipleMatchesError()";
+          break;
+        default:
+          error = "IlsError()";
+      }
+      throw new Error(error);
+      // TODO: catch HTTP timout errors
+      // These errors are returned when the SOAP client times out
+      // when connecting and when requesting (respectively).
+    }
+  }
+
+  search(query) {
+    return this.callAndFailGracefully({
+      type: "search_patrons",
+      message: {
+        search_string: query,
+        username: process.env["ILS_USERNAME"],
+        password: process.env["ILS_PASSWORD"],
+      },
+    });
+  }
+
+  create_patron(patron) {
+    if (!patron.validForIls) {
+      throw new Error("IlsError");
+    }
+    let ilsPatron = this.patronFields(patron);
+
+    return this.callAndFailGracefully({
+      type: "create_patron",
+      message: {
+        patron: ilsPatron,
+        username: process.env["ILS_USERNAME"],
+        password: process.env["ILS_PASSWORD"],
+      },
+    });
+  }
+
+  updatePatron(patron) {
+    if (!patron.validForIls) {
+      throw new Error("IlsError");
+    }
+    let ilsPatron = this.patronFields(patron);
+
+    return this.callAndFailGracefully({
+      type: "update_patron",
+      message: {
+        patron: ilsPatron,
+        username: process.env["ILS_USERNAME"],
+        password: process.env["ILS_PASSWORD"],
+      },
+    });
+  }
+
+  getPatronIdFromResponse(response) {
+    return response.body["multi_ref"][0]["patronId"];
+  }
+
+  available(barcodeOrUsername) {
+    const result = search(
+      `//${IlsHelper.BARCODE_FIELDTAG}//${barcodeOrUsername}`
+    );
+    return !!(result === "NotFoundError");
+  }
+
+  // Format a patron's address as ILS expects it.
+  formatAddress(address) {
+    return address.toUpperCase.gsub(/[\n\r]+/, "$");
+  }
+
+  // Internal: Format card fields for ILS patron creation/updates.
+  formattedPatronData(patron) {
+    let email;
+    let birthdate;
+    let workAddress;
+    if (patron.email) {
+      email = patron.email.gsub(/\s+/, "").toUpperCase;
+    }
+    if (patron.birthdate) {
+      // birthdate = patron.birthdate.strftime("%Y%m%d000000");
```

Comment:
Was this line supposed to be removed...?

Reply:
Right, so many of this is still the ruby to js conversion and, like the description says, it's skeleton code without updating too much of the logic. In this case, the string needs to be converted but it does it a few times so it might be better as a utility function which I haven't gotten to yet.

Reply:
I'll update this whole file in a different PR since it is its own ticket. This file doesn't affect the other files since it is mocked in tests.